File: leetcode/prefix_sum/560. Subarray Sum Equals K/js/main.js
// Hashmap solution
// Time O(N)
// Space O(N)
const subarraySum = (nums, k) => {
  let n = nums.length;
  let map = { 0: 1 };
  let sum = 0;
  let cnt = 0;

  for (let i = 0; i < n; i++) {
    sum += nums[i];

    if (map[sum - k] > 0) {
      cnt += map[sum - k];
    }

    map[sum] = ~~map[sum] + 1;
  }

  return cnt;
};

File: leetcode/prefix_sum/560. Subarray Sum Equals K/py/main.py
class Solution:
    # Time O(N)
    # Space O(N)
    def subarraySum(self, nums: List[int], k: int) -> int:
        mapping = {0: 1}
        current = 0
        n = len(nums)
        ans = 0

        for i in range(n):
            current += nums[i]

            if current - k in mapping:
                ans += mapping.get(current - k)

            mapping[current] = mapping.get(current, 0) + 1

        return ans
